name: Lint

on: [push, pull_request, workflow_dispatch]

permissions:
  contents: read

env:
  FORCE_COLOR: 1
  RUFF_FORMAT: github

concurrency:
  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
  cancel-in-progress: true

jobs:
  lint:
    runs-on: ubuntu-latest
    timeout-minutes: 10

    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v4
        with:
          python-version: "3.x"
      - uses: pre-commit/action@v3.0.0

  patchcheck:
    if: github.event_name == 'pull_request'
    runs-on: ubuntu-latest
    timeout-minutes: 10

    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 2
          show-progress: false
      # Adapted from https://github.com/actions/checkout/issues/520#issuecomment-1167205721
      - name: Fetch merge-base
        run: >
          git fetch
          origin
          +${{ github.event.pull_request.head.sha }}:remotes/origin/${{ github.event.pull_request.head.ref }}
          --depth=$(( ${{ github.event.pull_request.commits }} + 1 ))
          --no-tags
          --prune
          --no-recurse-submodules
      - uses: actions/setup-python@v4
        with:
          python-version: "3.x"
      - name: "Run patchcheck.py"
        run: python ./Tools/patchcheck/patchcheck.py --ci true
